litex_boards/platforms/sipeed_tang_nano_4k.py

```diff
@@ -91,7 +91,15 @@ def __init__(self, toolchain="gowin"):
         GowinPlatform.__init__(self, "GW1NSR-LV4CQN48PC6/I5", _io, _connectors, toolchain=toolchain, devicename="GW1NSR-4C")
         # Licensed GoWin IDE v. 1.9.7.06 Beta
         # GowinPlatform.__init__(self, "GW1NSR-LV4CQN48PC7/I6", _io, _connectors, toolchain=toolchain, devicename="GW1NSR-4C")
-        self.toolchain.options["use_mode_as_gpio"] = 1
+
+        if toolchain=="gowin":
+            # this is just a software check that doesn't translate to hardware
+            # apicula does not implement this check, so beware of bricking your board
+            self.toolchain.options["use_mode_as_gpio"] = 1
+        else:
+            # yosys doesn't know that 4K doesn't have lutram so we tell it
+            self.toolchain._synth_opts += " -nolutram"
+
```

Comment:
A last question here: maybe is it possible to move this part to apicula toolchain? This avoid to have to copy/paste these lines for all boards with this trick.

Reply:
It'd have to be spread across several tools I guess.

Yosys is currently oblivious to the specific device it's synthesizing for, so you could add a `-family` flag there or something but that doesn't really have any advantage over `-nolutram` until there are more significant differences.

We could add a `--mode-as-gpio` flag to the packer that implements the check, but like the comment said it doesn't translate to anything in hardware. It's more of a "are you sure you want to do this" thing. Given that you wrote the constraints file, I would assume the answer is "yes". All of the other `as-gpio` flags actually change bits to switch the functionality.